File: year2017/day2/d2.pl
#!/usr/bin/env perl

use strict;
use warnings;

my $sum_1 = 0;
my $sum_2 = 0;

foreach my $line (<STDIN>) {
  my @row = sort { $a <=> $b } split(" ", $line);
  $sum_1 += $row[-1] - $row[0];

  my $found = 0;
  while (@row and !$found) {
    my $a = pop @row;

    foreach my $b (@row) {
      if ($a % $b == 0) {
        $sum_2 += $a / $b;
        $found = 1;
      }
    }
  }
}

print "$sum_1\n";
print "$sum_2\n";

File: year2017/day2/d2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>


int main() {

    int sum_1 = 0;
    int sum_2 = 0;
    std::string line;
    std::vector< std::vector<int> > matrix;

    while ( getline( std::cin, line ) ) {
      std::istringstream in( line );
      matrix.push_back(
            std::vector<int>(
              std::istream_iterator<int>(in), std::istream_iterator<int>() )
            );
    }

    for (std::size_t i = 0; i < matrix.size(); i++) {
      std::sort(matrix[i].begin(), matrix[i].end());
      sum_1 += matrix[i].back() - matrix[i].front();

      bool found = false;
      while (!matrix[i].empty() && !found) {
        int last = matrix[i].back();
        matrix[i].pop_back();

        for (std::size_t j = 0; j < matrix[i].size(); j++) {
          if (last % matrix[i][j] == 0) {
            sum_2 += last / matrix[i][j];
            found = true;
          }
        }
      }
    }

    std::cout << sum_1 << "\n";
    std::cout << sum_2 << "\n";
}
